core/deployment/src/main/java/io/quarkus/deployment/ExtensionLoader.java

```diff
@@ -214,7 +218,45 @@ public boolean canHandleObject(final Object obj, final boolean staticInit) {
                 .produces(RunTimeConfigurationProxyBuildItem.class)
                 .produces(BytecodeRecorderObjectLoaderBuildItem.class)
                 .build());
-        return result;
+
+        return result.andThen(provideCapabilities(classLoader, appModel, readResult, launchMode, devModeType));
+    }
+
+    private static Consumer<BuildChainBuilder> provideCapabilities(ClassLoader classLoader, AppModel appModel,
+            BuildTimeConfigurationReader.ReadResult readResult, LaunchMode launchMode, DevModeType devModeType) {
+        Map<Class<? extends BooleanSupplier>, BooleanSupplier> condCache = new HashMap<>();
+
+        Consumer<BuildChainBuilder> stepConfig = Functions.discardingConsumer();
+        BooleanSupplier alwaysAddStep = () -> true;
+        for (CapabilityContract contract : appModel.getCapabilityContracts().values()) {
+            final String provider = contract.getExtension();
+            for (String capability : contract.getProvidesCapabilities()) {
+                int conditionIndex = capability.indexOf('?');
+                final String name = conditionIndex < 0 ? capability : capability.substring(0, conditionIndex);
+                BooleanSupplier buildIf = alwaysAddStep;
+                int testClassStart;
+                while (conditionIndex > 0) {
+                    final boolean inv = conditionIndex < capability.length() - 1
+                            && capability.charAt(conditionIndex + 1) == '!';
+                    testClassStart = conditionIndex + (inv ? 2 : 1);
+                    conditionIndex = capability.indexOf('?', testClassStart + 1);
+                    final String testClassName = capability
+                            .substring(testClassStart, conditionIndex > 0 ? conditionIndex : capability.length());
+                    Class<? extends BooleanSupplier> testClass;
+                    try {
+                        testClass = (Class<? extends BooleanSupplier>) classLoader.loadClass(testClassName);
```

Comment:
This kind of cast can be made safe (and more easily debuggable as a result):

```suggestion
                        testClass = classLoader.loadClass(testClassName).asSubclass(BooleanSupplier.class);
```